framework/include/utils/PerfGraph.h

```diff
@@ -0,0 +1,109 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#ifndef PERFGRAPH_H
+#define PERFGRAPH_H
+
+// MOOSE Includes
+#include "MooseTypes.h"
+#include "PerfNode.h"
+#include "PerfGuard.h"
+
+// System Includes
+#include <array>
+
+#define MAX_STACK_SIZE 100
+
+/**
+ * The PerfGraph will hold the master list of all registered performance segments and
+ * the head PerfNode
+ */
+class PerfGraph
+{
+public:
+  /**
+   * Create a new PerfGraph
+   */
+  PerfGraph();
+
+  /**
+   * Destructor
+   */
+  ~PerfGraph();
+
+  /**
+   * Registers a named section of code
+   *
+   * @return The unique ID to use for that section
+   */
+  PerfID registerSection(std::string section_name);
+
+  /**
+   * Add a Node onto the end of the end of the current callstack
+   */
+  void push(PerfID id);
+
+  /**
+   * Remove a Node from the end of the current scope
+   *
+   * @duration The amount of time the Node took
+   */
+  void pop(std::chrono::steady_clock::duration duration);
+
+  /**
+   * Print the tree out
+   */
+  void print();
```

Comment:
I suggest added in an ostream parameter so we can control where this gets printed.

Reply:
👍